scripts/tests/cross-test-ci-all.sh

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+# all versions to use for testing
+versions=("v0.2.1")
+>&2 echo "Running backwards-compatibility tests for versions: ${versions[*]}"
+
+# signal to downstream test scripts
+export FM_BACKWARDS_COMPATIBILITY_TEST=1
+
+function nix_build_binary_for_version() {
+  binary="$1"
+  version="$2"
+  echo "$(nix build 'github:fedimint/fedimint/'"$version"'#'"$binary" --no-link --print-out-paths)/bin/$binary"
+}
+
+function use_fed_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_FEDIMINTD_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling fed binaries for version $version..."
+    FM_FEDIMINTD_BASE_EXECUTABLE="$(nix_build_binary_for_version 'fedimintd' "$version")"
+    export FM_FEDIMINTD_BASE_EXECUTABLE
+  fi
+}
+
+function use_client_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_FEDIMINT_CLI_BASE_EXECUTABLE
+    unset FM_GATEWAY_CLI_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling client binaries for version $version..."
+    FM_FEDIMINT_CLI_BASE_EXECUTABLE="$(nix_build_binary_for_version 'fedimint-cli' "$version")"
+    export FM_FEDIMINT_CLI_BASE_EXECUTABLE
+    FM_GATEWAY_CLI_BASE_EXECUTABLE="$(nix_build_binary_for_version 'gateway-cli' "$version")"
+    export FM_GATEWAY_CLI_BASE_EXECUTABLE
+  fi
+}
+
+function use_gateway_binaries_for_version() {
+  version=$1
+  if [[ "$version" == "current" ]]; then
+    unset FM_GATEWAYD_BASE_EXECUTABLE
+  else
+    >&2 echo "Compiling gateway binaries for version $version..."
+    FM_GATEWAYD_BASE_EXECUTABLE="$(nix_build_binary_for_version 'gatewayd' "$version")"
+    export FM_GATEWAYD_BASE_EXECUTABLE
+  fi
+}
+
+test_results="fed_version,client_version,gateway_version,exit_code\n"
+has_failure=false
+versions+=("current")
+for fed_version in "${versions[@]}"; do
+  for client_version in "${versions[@]}"; do
+    for gateway_version in "${versions[@]}"; do
```

Comment:
During the deep dive we discussed `fedimintd` and `gatewayd` will be on the same version since we should assume gateway operators are technically savvy. I'm hesitant to make the assumption that all gateway operators will coordinate upgrades with all federations they're connected to and think it would be good to know if we're introducing breaking changes.

I'm open to removing if others feel strongly.

Reply:
Happy to cover it in tests, more is better, but we'll have to decide what to run when. The full matrix will likely be too slow for running on every push to a PR.

Reply:
Instead of one `fedimint-cli x fedimintd x gateways`, I'd rather do targeted ``fedimint-cli x fedimintd`, `fedimintd x gateways`, `fedimint-cli x gateways`, and each only with tests that make sense for it.

There might be some cases where "x can't communicate with y, but only if z is in a weird version", but seems significantly less likely and we could run a full matrix before releases.

Reply:
I agree that sounds like a good improvement. I put up this PR with the following feedback in mind.

> This is probably the most important PR unless severe bugs are found, otherwise we won't be able to keep developing with confidence.

https://github.com/fedimint/fedimint/pull/3982#issuecomment-1873857758

Do you think it's okay to address in followup PRs or is this a blocker for merging? For context, I won't have much bandwidth this week to audit which tests make sense specifically for backwards compatibility, so this will delay merging any backwards compatibility tests for a while.